msb/camera/msb_camera.py

```diff
@@ -1,78 +1,109 @@
+from msb.config import load_config
+
+timeformat = "%Y-%m-%d %X"
+color = (0, 255, 0)
+origin = (0, 30)
+font = cv2.FONT_HERSHEY_SIMPLEX
+scale = 1
+thickness = 2
 
-from msb.camera.CameraConfig import CameraConfig
 
 def signal_handler(sig, frame):
-    print("msb_piletrack.py exit")
+    print("msb_camera.py exit")
     sys.exit(0)
 
+
 def get_datetime() -> datetime.datetime:
     return datetime.datetime.fromtimestamp(time.time(), tz=datetime.timezone.utc)
 
-def get_datetime_str(timeformat : str = "%Y%m%dT%H%M%S.%f%z", timestamp=None) -> str:
-    if timestamp:
-        return timestamp.strftime(timeformat)
-    else:
-        return get_datetime().strftime(timeformat)
-
-def exit_handler(sig, frame):
-    global camera
-    camera.stop_recording()
-    print(f"test_camera.py received signal {sig}, exiting")
-    sys.exit()
-
-def get_new_fhandle(ts : datetime.datetime, folder : str = 'curdir', prefix : str = "msb-video") -> str:
-    if folder == 'curdir':
-        output_dir = os.path.abspath(os.path.curdir)
-    else:
-        output_dir = folder
-        if not os.path.isdir(output_dir):
-            raise Exception(f"no such file or directory: {folder}")
+
+def get_datetime_str(timestamp: datetime.datetime, config: CameraConf) -> str:
+    return timestamp.strftime(config.timeformat)
+
+
+def get_new_fhandle(timestamp: datetime.datetime, config: CameraConf) -> str:
+    ts_str = get_datetime_str(timestamp, config)
+    if not os.path.isdir(config.video_dir):
+        try:
+            os.makedirs(config.video_dir, exists_ok=True)
+        except Exception as e:
+            print(f"failed to create output directory: {config.video_dir}")
             sys.exit()
-    ts_str = get_datetime_str(timestamp=ts)
-    return os.path.join(output_dir, f"{prefix}_{ts_str}.h264")
+    return os.path.join(config.video_dir, f"{config.serial_number}_{ts_str}.h264")
+
+
+def apply_timestamp(request):
+    global timeformat
+    timestamp = time.strftime(timeformat)
+    with MappedArray(request, "main") as m:
+        cv2.putText(m.array, timestamp, origin, font, scale, color, thickness)
+
 
-def setup_camera(camera_config : CameraConfig):
+def setup_camera(config: CameraConf):
     """
     setup raspberry pi camera for recording
-    """ 
-    camera = picamera.PiCamera()
-    camera.resolution = (camera_config.width, camera_config.height)
-    camera.framerate = camera_config.fps
-    camera.annotate_background = picamera.Color('black')
-    camera.annotate_text = get_datetime_str()
-    camera.annotate_text_size = 16
+    """
+    camera = Picamera2()
+    camera.configure(
+        camera.create_video_configuration(main={"size": (config.width, config.height)})
+    )
+    camera.video_configuration.controls.FrameRate = config.fps
+    camera.pre_callback = apply_timestamp
     return camera
 
-def msb_camera(camera_config : CameraConfig):
-    if camera_config.verbose:
+
+def setup_globals(config: CameraConf):
```

Comment:
When using a class, this could be done in the constructor.

Reply:
good point, see previous comment with regards to the usage of global variables for the timestamp function.